File: portal/lib/analytics.js
import { getRedis, KEYS } from "./redis.js";

export const BOT_BUCKETS = [
  "Glama",
  "Claude/Anthropic",
  "OpenAI",
  "PythonScrapers",
  "ScriptClients",
  "Human",
  "UnknownBots",
];

const ZSET_UA = "beacon:stats:top_ua";
const ZSET_IP = "beacon:stats:top_ip";
const PREFIX_BOT = "beacon:stats:bot:";
const KEY_HUMAN_VISITS = "beacon:stats:human_visits";
const KEY_BOT_VISITS = "beacon:stats:bot_visits";
const KEY_HUMAN_API = "beacon:stats:human_api_hits";
const KEY_BOT_API = "beacon:stats:bot_api_hits";

export function clientIp(req) {
  const xf = req.headers["x-forwarded-for"] || req.headers["x-real-ip"] || "";
  const ip = String(xf).split(",")[0].trim();
  return ip || "unknown";
}

export function classifyClient(ua, event) {
  const u = (ua || "").toLowerCase();

  if (u.includes("glama")) return "Glama";
  if (u.includes("claudebot") || u.includes("anthropic")) return "Claude/Anthropic";
  if (u.includes("gptbot") || u.includes("oai-searchbot") || u.includes("chatgpt-user")) {
    return "OpenAI";
  }
  if (
    u.includes("python-requests") ||
    u.includes("httpx") ||
    u.includes("aiohttp") ||
    u.includes("urllib")
  ) {
    return "PythonScrapers";
  }
  if (
    u.includes("axios") ||
    u.includes("node") ||
    u.includes("undici") ||
    u.includes("curl") ||
    u.includes("wget") ||
    u.includes("go-http")
  ) {
    return "ScriptClients";
  }
  if (u.includes("bot") || u.includes("crawler") || u.includes("spider") || u.includes("scan")) {
    return "UnknownBots";
  }
  if (
    event === "visit" ||
    u.includes("mozilla") ||
    u.includes("chrome") ||
    u.includes("safari") ||
    u.includes("firefox") ||
    u.includes("edg/")
  ) {
    return "Human";
  }
  return "UnknownBots";
}

function uaLabel(ua) {
  const s = String(ua || "unknown").slice(0, 200);
  return s || "unknown";
}

export async function recordEvent(event, req) {
  const client = getRedis();
  if (!client) {
    return { ok: false, error: "redis_not_configured" };
  }

  const metricKey = KEYS[event];
  if (!metricKey) {
    return { ok: false, error: "invalid_event" };
  }

  const ua = req.headers["user-agent"] || "unknown";
  const ip = clientIp(req);
  const category = classifyClient(ua, event);
  const uaKey = uaLabel(ua);

  const pipe = client.pipeline();
  pipe.incr(metricKey);
  pipe.incr(`${PREFIX_BOT}${category}`);
  pipe.zincrby(ZSET_UA, 1, uaKey);
  pipe.zincrby(ZSET_IP, 1, ip);

  if (event === "visit") {
    if (category === "Human") pipe.incr(KEY_HUMAN_VISITS);
    else pipe.incr(KEY_BOT_VISITS);
  }
  if (event === "api_hit") {
    if (category === "Human") pipe.incr(KEY_HUMAN_API);
    else pipe.incr(KEY_BOT_API);
  }

  const results = await pipe.exec();
  const total = Number(results?.[0]) || 0;

  return { ok: true, event, total, category, ip: ip.slice(0, 8) + "…" };
}

export async function getPublicStats() {
  const client = getRedis();
  if (!client) {
    return { ok: true, configured: false, visits: 0, api_hits: 0, installs: 0 };
  }

  const [visits, api_hits, installs] = await client.mget(
    KEYS.visits,
    KEYS.api_hits,
    KEYS.installs,
  );

  return {
    ok: true,
    configured: true,
    visits: Number(visits) || 0,
    api_hits: Number(api_hits) || 0,
    installs: Number(installs) || 0,
  };
}

async function topFromZset(client, key, limit = 15) {
  const rows = await client.zrange(key, 0, limit - 1, { rev: true, withScores: true });
  if (!rows?.length) return [];

  const out = [];
  for (let i = 0; i < rows.length; i += 2) {
    out.push({ value: rows[i], hits: Number(rows[i + 1]) || 0 });
  }
  return out;
}

export async function getAnalyticsDetails() {
  const client = getRedis();
  if (!client) {
    return {
      ok: false,
      configured: false,
      error: "redis_not_configured",
    };
  }

  const keys = [
    KEYS.visits,
    KEYS.api_hits,
    KEYS.installs,
    KEY_HUMAN_VISITS,
    KEY_BOT_VISITS,
    KEY_HUMAN_API,
    KEY_BOT_API,
    ...BOT_BUCKETS.map((b) => `${PREFIX_BOT}${b}`),
  ];

  const values = await client.mget(...keys);
  const pick = (i) => Number(values[i]) || 0;

  const bot_traffic_breakdown = {};
  BOT_BUCKETS.forEach((bucket, i) => {
    const v = pick(7 + i);
    if (v > 0) bot_traffic_breakdown[bucket] = v;
  });

  const topUa = await topFromZset(client, ZSET_UA, 20);
  const topIp = await topFromZset(client, ZSET_IP, 15);

  const top_consumers = topUa.map((row) => ({
    user_agent: row.value,
    hits: row.hits,
    category: classifyClient(row.value, "api_hit"),
  }));

  const top_ips = topIp.map((row) => ({
    ip: row.value,
    hits: row.hits,
  }));

  return {
    ok: true,
    configured: true,
    summary: {
      total_visits: pick(0),
      api_hits: pick(1),
      npm_installs: pick(2),
      human_browser_visits: pick(3),
      bot_browser_visits: pick(4),
      human_api_hits: pick(5),
      bot_api_hits: pick(6),
    },
    bot_traffic_breakdown,
    top_consumers,
    top_ips,
    generated_at: new Date().toISOString(),
  };
}

File: registry/analytics.py
"""
Request analytics — SQLite-backed access log and admin rollups for the GCP registry.
"""

from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Optional

import turso

ADMIN_SECRET = os.environ.get("ADMIN_SECRET_PASSWORD", "")

# Paths we skip (noise / static)
_SKIP_PREFIXES = ("/docs", "/openapi.json", "/redoc", "/favicon")


def _utcnow() -> str:
    return datetime.now(timezone.utc).isoformat()


def _client_ip(request) -> str:
    xf = request.headers.get("x-forwarded-for") or request.headers.get("x-real-ip") or ""
    ip = str(xf).split(",")[0].strip()
    if request.client and not ip:
        ip = request.client.host or ""
    return ip or "unknown"


def _traffic_bucket(ua: str) -> str:
    u = (ua or "").lower()
    if "glama" in u:
        return "Glama"
    if "claudebot" in u or "anthropic" in u:
        return "Claude/Anthropic"
    if "gptbot" in u or "oai-searchbot" in u or "chatgpt" in u:
        return "OpenAI"
    if "perplexity" in u:
        return "Perplexity"
    if "python-requests" in u or "httpx" in u or "aiohttp" in u or "urllib" in u:
        return "PythonScrapers"
    if "beacon-indexer" in u or "scraper" in u:
        return "BeaconScraper"
    if any(x in u for x in ("axios", "node", "undici", "curl", "wget", "go-http")):
        return "ScriptClients"
    if any(x in u for x in ("beacon", "mcp", "cursor", "cline", "windsurf")):
        return "AgentClients"
    if any(x in u for x in ("bot", "crawler", "spider", "bytespider", "bingbot")):
        return "UnknownBots"
    if any(x in u for x in ("mozilla", "chrome", "safari", "firefox", "edg/")):
        return "Human"
    return "Other"


def verify_admin(request) -> bool:
    if not ADMIN_SECRET:
        return False
    auth = (request.headers.get("authorization") or "").strip()
    if auth.lower().startswith("bearer "):
        return auth[7:].strip() == ADMIN_SECRET
    token = request.headers.get("x-admin-token")
    return isinstance(token, str) and token.strip() == ADMIN_SECRET


def ensure_analytics_schema() -> None:
    turso.execute(
        """
        CREATE TABLE IF NOT EXISTS access_log (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            ts TEXT NOT NULL,
            method TEXT NOT NULL,
            path TEXT NOT NULL,
            client_class TEXT NOT NULL,
            client_bucket TEXT NOT NULL,
            ip TEXT,
            user_agent TEXT
        )
        """
    )
    turso.execute("CREATE INDEX IF NOT EXISTS idx_access_log_ts ON access_log(ts DESC)")
    turso.execute("CREATE INDEX IF NOT EXISTS idx_access_log_bucket ON access_log(client_bucket)")
    turso.execute(
        """
        CREATE TABLE IF NOT EXISTS access_totals (
            key TEXT PRIMARY KEY,
            value INTEGER NOT NULL DEFAULT 0
        )
        """
    )


def log_request(request, client_class: str) -> None:
    path = request.url.path
    if not path.startswith("/api/"):
        return
    if any(path.startswith(p) for p in _SKIP_PREFIXES):
        return
    if path.startswith("/api/admin"):
        return

    ua = (request.headers.get("user-agent") or "unknown")[:512]
    bucket = _traffic_bucket(ua)
    ip = _client_ip(request)

    try:
        turso.execute(
            """
            INSERT INTO access_log (ts, method, path, client_class, client_bucket, ip, user_agent)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            [_utcnow(), request.method, path, client_class, bucket, ip, ua],
        )
        turso.execute(
            """
            INSERT INTO access_totals (key, value) VALUES (?, 1)
            ON CONFLICT(key) DO UPDATE SET value = value + 1
            """,
            [f"reads:{bucket}"],
        )
        turso.execute(
            """
            INSERT INTO access_totals (key, value) VALUES ('api_reads_total', 1)
            ON CONFLICT(key) DO UPDATE SET value = value + 1
            """
        )
    except turso.TursoError:
        pass


def count_active_agents(pushed_recently_fn) -> int:
    rows = turso.execute(
        "SELECT pushed_at FROM agents WHERE is_fraudulent = 0 AND pushed_at IS NOT NULL"
    )
    return sum(1 for r in rows if pushed_recently_fn(r.get("pushed_at")))


def get_admin_analytics(pushed_recently_fn) -> dict:
    ensure_analytics_schema()

    total = int(turso.execute("SELECT COUNT(*) AS c FROM agents")[0]["c"])
    scraped = int(
        turso.execute(
            "SELECT COUNT(*) AS c FROM agents WHERE registration_source = 'scraper'"
        )[0]["c"]
    )
    sdk = int(
        turso.execute(
            "SELECT COUNT(*) AS c FROM agents WHERE registration_source = 'sdk'"
        )[0]["c"]
    )
    flagged = int(
        turso.execute("SELECT COUNT(*) AS c FROM agents WHERE is_fraudulent = 1")[0]["c"]
    )
    active = count_active_agents(pushed_recently_fn)

    totals_rows = turso.execute("SELECT key, value FROM access_totals")
    totals_map = {r["key"]: int(r["value"]) for r in totals_rows}
    api_reads = totals_map.get("api_reads_total", 0)

    log_count = int(turso.execute("SELECT COUNT(*) AS c FROM access_log")[0]["c"])

    bucket_rows = turso.execute(
        """
        SELECT client_bucket AS bucket, COUNT(*) AS hits
          FROM access_log
         GROUP BY client_bucket
         ORDER BY hits DESC
         LIMIT 20
        """
    )
    bot_traffic_breakdown = {r["bucket"]: int(r["hits"]) for r in bucket_rows}

    ua_rows = turso.execute(
        """
        SELECT user_agent, client_bucket, COUNT(*) AS hits
          FROM access_log
         GROUP BY user_agent, client_bucket
         ORDER BY hits DESC
         LIMIT 25
        """
    )
    top_consumers = [
        {
            "user_agent": (r["user_agent"] or "")[:200],
            "category": r["client_bucket"],
            "hits": int(r["hits"]),
        }
        for r in ua_rows
    ]

    ip_rows = turso.execute(
        """
        SELECT ip, COUNT(*) AS hits
          FROM access_log
         GROUP BY ip
         ORDER BY hits DESC
         LIMIT 20
        """
    )
    top_ips = [{"ip": r["ip"], "hits": int(r["hits"])} for r in ip_rows]

    path_rows = turso.execute(
        """
        SELECT path, COUNT(*) AS hits
          FROM access_log
         GROUP BY path
         ORDER BY hits DESC
         LIMIT 15
        """
    )
    top_endpoints = [{"path": r["path"], "hits": int(r["hits"])} for r in path_rows]

    human_reads = sum(
        v for k, v in totals_map.items() if k.startswith("reads:") and "Human" in k
    ) or bot_traffic_breakdown.get("Human", 0)
    bot_reads = api_reads - human_reads if api_reads > human_reads else sum(
        v for b, v in bot_traffic_breakdown.items() if b != "Human"
    )

    return {
        "ok": True,
        "authenticated": True,
        "generated_at": _utcnow(),
        "storage": "sqlite",
        "storage_path": turso.DB_PATH,
        "summary": {
            "total_agents_indexed": total,
            "agents_harvested_by_scraper": scraped,
            "agents_registered_via_sdk": sdk,
            "active_agents_90d": active,
            "flagged_agents": flagged,
            "total_api_reads_logged": api_reads,
            "human_api_reads": human_reads,
            "bot_api_reads": bot_reads,
            "access_log_rows": log_count,
        },
        "bot_traffic_breakdown": bot_traffic_breakdown,
        "top_consumers": top_consumers,
        "top_ips": top_ips,
        "top_endpoints": top_endpoints,
    }
